tests/datastore/testthreadproxyds.nim

```diff
@@ -0,0 +1,80 @@
+import std/options
+import std/sequtils
+import std/os
+import std/cpuinfo
+import std/algorithm
+import std/importutils
+
+import pkg/asynctest
+import pkg/chronos
+import pkg/stew/results
+import pkg/stew/byteutils
+import pkg/taskpools
+import pkg/questionable/results
+
+import pkg/datastore/sql
+import pkg/datastore/fsds
+import pkg/datastore/threads/threadproxyds {.all.}
+
+import ./dscommontests
+import ./querycommontests
+
+suite "Test Basic ThreadDatastore with SQLite":
+
+  var
+    sqlStore: Datastore
+    ds: ThreadDatastore
+    taskPool: Taskpool
+    key = Key.init("/a/b").tryGet()
+    bytes = "some bytes".toBytes
+    otherBytes = "some other bytes".toBytes
+
+  setupAll:
+    sqlStore = SQLiteDatastore.new(Memory).tryGet()
+    taskPool = Taskpool.new(countProcessors() * 2)
+    ds = ThreadDatastore.new(sqlStore, taskPool).tryGet()
+
+  teardownAll:
+    (await ds.close()).tryGet()
+    taskPool.shutdown()
+
+  basicStoreTests(ds, key, bytes, otherBytes)
+
+
+suite "Test Query ThreadDatastore with SQLite":
+
+  var
+    sqlStore: Datastore
+    ds: ThreadDatastore
+    taskPool: Taskpool
+    key = Key.init("/a/b").tryGet()
+    bytes = "some bytes".toBytes
+    otherBytes = "some other bytes".toBytes
+
+  setup:
+    sqlStore = SQLiteDatastore.new(Memory).tryGet()
+    taskPool = Taskpool.new(countProcessors() * 2)
+    ds = ThreadDatastore.new(sqlStore, taskPool).tryGet()
+
+  teardown:
+    (await ds.close()).tryGet()
+    taskPool.shutdown()
+
+  queryTests(ds, true)
+
+# TODO: needs a per key lock to work
```

Comment:
Yah, I think it'd be possible to do something like my `signalpool` for the key-lock. It's pretty much the same pattern just without pre-allocating the keys and doing the await on whether the key is used.

Reply:
It's actually much simpler, we don't need to do this synchronization at the thread level, same as with the semaphore, I realized that after I wrote the threaded semaphore. We just need a table where we store a future (any future), for the that the async `get/set/has/put/del` check and set, before spawning the task. Since we have an SPMC flow, we can do away with thread level locking in most places, this will also serialize read/write operations since the "locking" will happen at the future level and not at the thread level. 

I'll integrate this shortly, it sounds like a relatively straightforward thing to do.